File: src/infrastructure/securityutils/PasswordHashProvider.cs
using System.Security.Cryptography;
using System.Text;
using core.fs.Shared.Adapters.Authentication;

namespace securityutils;

// https://dotnetcodr.com/2017/10/26/how-to-hash-passwords-with-a-salt-in-net-2/
public class PasswordHashProvider : IPasswordHashProvider
{
    public HashAndSalt GenerateHashAndSalt(string password, int saltLength)
    {
        return Generate(
            Encoding.UTF8.GetBytes(password),
            RandomNumberGenerator.GetBytes(saltLength)
        );
    }

    public string GenerateHash(string password, string salt)
    {
        var result = Generate(
            Encoding.UTF8.GetBytes(password),
            Convert.FromBase64String(salt)
        );

        return result.Hash;
    }

    private static HashAndSalt Generate(byte[] passwordAsBytes, byte[] saltBytes)
    {
        var passwordWithSaltBytes = new List<byte>();
        passwordWithSaltBytes.AddRange(passwordAsBytes);
        passwordWithSaltBytes.AddRange(saltBytes);
        byte[] hashBytes = SHA512.Create().ComputeHash(passwordWithSaltBytes.ToArray());
        return new HashAndSalt(hash: Convert.ToBase64String(hashBytes), salt: Convert.ToBase64String(saltBytes));
    }
}

File: src/infrastructure/securityutils.fs/PasswordHashProvider.fs
namespace securityutils

open System
open System.Security.Cryptography
open System.Text
open core.fs.Adapters.Authentication

// https://dotnetcodr.com/2017/10/26/how-to-hash-passwords-with-a-salt-in-net-2/
type PasswordHashProvider() =
    
    let generate (passwordBytes: byte[]) (saltBytes: byte[]) : HashAndSalt =
        let passwordWithSaltBytes = Array.concat [passwordBytes; saltBytes]
        use sha512 = SHA512.Create()
        let hashBytes = sha512.ComputeHash(passwordWithSaltBytes)
        { Hash = Convert.ToBase64String(hashBytes)
          Salt = Convert.ToBase64String(saltBytes) }
    
    member _.GenerateHashAndSalt(password: string, saltLength: int) : HashAndSalt =
        let passwordBytes = Encoding.UTF8.GetBytes(password)
        let saltBytes = RandomNumberGenerator.GetBytes(saltLength)
        generate passwordBytes saltBytes
    
    member _.GenerateHash(password: string, salt: string) : string =
        let passwordBytes = Encoding.UTF8.GetBytes(password)
        let saltBytes = Convert.FromBase64String(salt)
        let result = generate passwordBytes saltBytes
        result.Hash
    
    interface IPasswordHashProvider with
        member this.GenerateHashAndSalt password saltLength =
            this.GenerateHashAndSalt(password, saltLength)
        
        member this.GenerateHash password salt =
            this.GenerateHash(password, salt)

